Rules/AvoidOverwritingBuiltInCmdlets.cs

```diff
@@ -0,0 +1,264 @@
+                foreach (KeyValuePair<string, HashSet<string>> cmdletSet in _cmdletMap)
+                {
+                    if (cmdletSet.Value.Contains(functionName))
+                    {
+                        diagnosticRecords.Add(CreateDiagnosticRecord(functionName, cmdletSet.Key, functionDef.Extent));
+                    }
+                }
+            }
+
+            return diagnosticRecords;
+        }
+
+
+        private DiagnosticRecord CreateDiagnosticRecord(string FunctionName, string PSVer, IScriptExtent ViolationExtent)
+        {
+            var record = new DiagnosticRecord(
+                string.Format(CultureInfo.CurrentCulture,
+                    string.Format(Strings.AvoidOverwritingBuiltInCmdletsError, FunctionName, PSVer)),
+                ViolationExtent,
+                GetName(),
+                GetDiagnosticSeverity(),
+                ViolationExtent.File,
+                null
+            );
+            return record;
+        }
+
+
+        private bool ContainsReferenceFile(string directory, string reference)
+        {
+            return File.Exists(Path.Combine(directory, reference + ".json"));
+        }
+
+
+        private void ProcessDirectory(string path, IEnumerable<string> acceptablePlatformSpecs)
+        {
+            foreach (var filePath in Directory.EnumerateFiles(path))
+            {
+                var extension = Path.GetExtension(filePath);
+                if (String.IsNullOrWhiteSpace(extension)
+                    || !extension.Equals(".json", StringComparison.OrdinalIgnoreCase))
+                {
+                    continue;
+                }
+
+                var fileNameWithoutExt = Path.GetFileNameWithoutExtension(filePath);
+                if (acceptablePlatformSpecs != null
+                    && !acceptablePlatformSpecs.Contains(fileNameWithoutExt, StringComparer.OrdinalIgnoreCase))
+                {
+                    continue;
+                }
+
+                if (_cmdletMap.Keys.Contains(fileNameWithoutExt))
+                {
+                    continue;
+                }
+
+                _cmdletMap.Add(fileNameWithoutExt, GetCmdletsFromData(JObject.Parse(File.ReadAllText(filePath))));
+            }
+        }
+
+
+        private HashSet<string> GetCmdletsFromData(dynamic deserializedObject)
+        {
+            var cmdlets = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+            dynamic modules = deserializedObject.Modules;
+            foreach (dynamic module in modules)
+            {
+                if (module.ExportedCommands == null)
+                {
+                    continue;
+                }
+
+                foreach (dynamic cmdlet in module.ExportedCommands)
+                {
+                    var name = cmdlet.Name as string;
+                    if (name == null)
+                    {
+                        name = cmdlet.Name.ToObject<string>();
```

Comment:
Can you explain this? Every object has a `.ToString()` method anyway? What is happening here?

Reply:
Ah, yes. I stole it from another rule. Changed.